app/src/main/java/iband/com/ibandplayerdemo/PlayerActivity.java

```diff
@@ -0,0 +1,440 @@
+            mSbPosition.setMax((int) duration);
+        }
+        if (!mIsUserChangePosition) {
+            mSbPosition.setProgress((int) currentPosition);
+
+            if (mLastPosition / 1000 != currentPosition / 1000) {
+                mLastPosition = currentPosition;
+                mTvPosition.setText(Utils.msToMinuteSec(currentPosition));
+            }
+        }
+    }
+
+    @Override
+    public void onPlayerBufferPositionUpdate(IBandPlayer player, long bufferPosition) {
+        mSbPosition.setSecondaryProgress((int) mPlayer.getBufferPosition());
+    }
+
+    @Override
+    public void onStreamError(IBandStream iBandStream, IBandError iBandError) {
+        toast(iBandError.getErrorMessage());
+    }
+
+
+    @Override
+    public void onPlayerPlayableChanged(IBandPlayer iBandPlayer, boolean isPlayable) {
+        if (!isPlayable) {
+            mViewPLayer.setVRMode(false);
+            setDisplayMode();
+            mBtnPlayPause.setVisibility(View.GONE);
+            mBtnVideoScale.setVisibility(View.GONE);
+            mBtnVariant.setEnabled(false);
+            mViewPLayer.setVisibility(View.GONE);
+        } else {
+            mBtnPlayPause.setVisibility(View.VISIBLE);
+
+        }
+    }
+
+    @Override
+    public void onPlayerReachEnd(IBandPlayer iBandPlayer) {
+        mProgressPlayer.setVisibility(View.GONE);
+        mBtnPlayPause.setImageResource(R.drawable.ic_player_play);
+    }
+
+    @Override
+    public void onPlayerVariantsCreated(IBandPlayer iBandPlayer, IBandPlayerVariant[] iBandPlayerVariants) {
+        mVariants = new ArrayList<VariantHolder>();
+        for (int i = 0; i < iBandPlayerVariants.length; i++) {
+            mVariants.add(0, new VariantHolder(i, iBandPlayerVariants[i]));
+        }
+        mVariants.add(new VariantHolder(-1, null));
+        mVariantAdapter = new VariantAdapter(PlayerActivity.this, mVariants);
+        mVariantAdapter.setSelected(mVariants.size() - 1);
+        mLvSettings.setVisibility(View.GONE);
+        mLvSettings.setAdapter(mVariantAdapter);
+        mBtnVariant.setEnabled(true);
+
+    }
+
+    @Override
+    public void onPlayerCurrentVariantChanged(IBandPlayer iBandPlayer, int variantIndex, int oldIndex) {
+        mVariantAdapter.setActive(variantIndex);
+        mVariantAdapter.notifyDataSetChanged();
+    }
+
+    @Override
+    public void onPlayerError(IBandPlayer iBandPlayer, IBandError iBandError) {
+        mProgressPlayer.setVisibility(View.GONE);
+        mProgressStream.setVisibility(View.GONE);
+        toast(iBandError.getErrorMessage());
+    }
+
+    @Override
+    public void onStreamStateChanged(IBandStream iBandStream, IBandStream.State state, IBandStream.State old) {
+        if (old == IBandStream.State.INITIALIZE) {
+            mProgressStream.setVisibility(View.GONE);
+        }
+        if (iBandStream == mStream) {
+            if (iBandStream.getState() == IBandStream.State.CLOSED) {
+                mLvSettings.setAdapter(null);
```

Comment:
Move those commands to an informative function such as `setupEmptyView()`

Reply:
done